File: python/pt_verify.py
#!/usr/bin/env python3
"""Independent dual of the p_T / B(3) attack.

Does not import pt_acf.py. Rebuilds the m=3 maximizer, every cap-243
survivor, famous and Cassaigne controls, VdW counts, and p <= diam+1
from first principles (desub_geometry + kernel_regular primitives only).
"""
from __future__ import annotations

import os
import sys
from collections import defaultdict

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from desub_geometry import (  # noqa: E402
    FAMOUS,
    find_cube,
    first_acf_len,
    morph_from_words,
    prefix_sums,
)
from kernel_regular import cassaigne_iterate, is_primitive  # noqa: E402


def decode_code(code: int):
    imgs = []
    t = code
    for _a in range(4):
        trip = []
        for _r in range(3):
            trip.append(t & 3)
            t >>= 2
        imgs.append(tuple(trip))
    return imgs


def sigma_from_code(code: int):
    imgs = decode_code(code)
    return morph_from_words(imgs, (0, 1, 2, 3)), imgs


def incidence_from_imgs(imgs):
    M = [[0] * 4 for _ in range(4)]
    for a, img in enumerate(imgs):
        for x in img:
            M[x][a] += 1
    return M


def T_from_imgs(imgs):
    return tuple(sum(img) for img in imgs)


def additive_at(w, n: int):
    S = prefix_sums(w)
    N = len(S) - 1
    weights = [S[i + n] - S[i] for i in range(0, N - n + 1)]
    vals = set(weights)
    L, H = min(weights), max(weights)
    ntrip = 0
    for i in range(0, N - 3 * n + 1):
        if weights[i] == weights[i + n] == weights[i + 2 * n]:
            ntrip += 1
    return {
        "p": len(vals),
        "L": L,
        "H": H,
        "diam": H - L,
        "holes": (H - L + 1) - len(vals),
        "ntrip": ntrip,
        "weights": weights,
    }


def vdW(w, n: int) -> dict:
    S = prefix_sums(w)
    N = len(S) - 1
    color = [S[i + n] - S[i] for i in range(0, N - n + 1)]
    by = defaultdict(list)
    for i, c in enumerate(color):
        by[c].append(i)
    n_ap_any = 0
    n_ap_diff_n = 0
    for _c, pos in by.items():
        st = set(pos)
        for i in pos:
            for d in range(1, (N - n - i) // 2 + 1):
                if i + d in st and i + 2 * d in st:
                    n_ap_any += 1
                    if d == n:
                        n_ap_diff_n += 1
                    break
    return {
        "n_colors": len(by),
        "n_starts_with_some_mono_3AP": n_ap_any,
        "n_starts_with_diff_exactly_n": n_ap_diff_n,
    }


def vtm(cap: int) -> list[int]:
    imgs = {0: (0, 1, 2), 1: (0, 2), 2: (1,)}
    w = [0]
    while len(w) < cap:
        nxt = []
        for a in w:
            nxt.extend(imgs[a])
        if len(nxt) <= len(w):
            break
        w = nxt
    return w[:cap]


def parse_scan(path: str) -> dict:
    d: dict = {"push": []}
    with open(path) as f:
        for line in f:
            parts = line.split()
            if not parts:
                continue
            if parts[0] == "push":
                d["push"].append(
                    {
                        "code": int(parts[1]),
                        "seed": int(parts[3]),
                        "acf": int(parts[5]),
                    }
                )
            elif len(parts) >= 2 and parts[0] not in ("hist_nonzero", "best_imgs"):
                try:
                    d[parts[0]] = int(parts[1])
                except ValueError:
                    d[parts[0]] = parts[1]
    return d


def charpoly_coeffs(M):
    """Characteristic polynomial of a 4x4 integer matrix via Faddeev–LeVerrier."""
    n = 4
    I = [[1 if i == j else 0 for j in range(n)] for i in range(n)]

    def matmul(A, B):
        C = [[0] * n for _ in range(n)]
        for i in range(n):
            for k in range(n):
                for j in range(n):
                    C[i][j] += A[i][k] * B[k][j]
        return C

    def trace(A):
        return sum(A[i][i] for i in range(n))

    c = [1]
    A = [row[:] for row in I]
    Mk = [row[:] for row in M]
    for k in range(1, n + 1):
        ck = -trace(Mk) // k
        c.append(ck)
        # M_{k} = M (M_{k-1} + c_k I); next Mk = M * (current_acc)
        acc = [[Mk[i][j] + ck * I[i][j] for j in range(n)] for i in range(n)]
        Mk = matmul(M, acc)
        A = acc
    return c  # [1, c1, c2, c3, c4] for t^4 + c1 t^3 + ...


def main():
    fails = 0
    scan_path = "data/pt_scan3.txt"
    scan = parse_scan(scan_path)
    expected = {
        "N": 16777216,
        "n_prol": 11468800,
        "n_prim": 6993126,
        "n_3sum_prol": 9992241,
        "n_tried_prim_3sum": 6116013,
        "n_surv_cap": 86,
        "n_surv_2187": 0,
        "best2_acf": 1647,
        "best2_code": 12507201,
        "best2_seed": 1,
        "SURV243": 86,
        "hist_p50": 15,
        "hist_p90": 29,
        "hist_p99": 56,
    }
    for k, v in expected.items():
        if scan.get(k) != v:
            print("FAIL scan field", k, scan.get(k), "expected", v)
            fails += 1
        else:
            print("OK scan", k, v)
    if len(scan["push"]) != 86:
        print("FAIL push lines", len(scan["push"]))
        fails += 1

    # Rebuild every survivor, all prolongable seeds, cap 2187.
    best = -1
    best_code = -1
    best_seed = -1
    n_surv2 = 0
    acf_mismatch = 0
    for rec in scan["push"]:
        sigma, imgs = sigma_from_code(rec["code"])
        M = incidence_from_imgs(imgs)
        if not is_primitive(M) or len(set(T_from_imgs(imgs))) < 3:
            print("FAIL survivor filter", rec)
            fails += 1
            continue
        local = -1
        lseed = -1
        for a in range(4):
            if imgs[a][0] != a:
                continue
            w = sigma.iterate(a, 2187)
            L = first_acf_len(w)
            if L > local:
                local = L
                lseed = a
        if local >= 2187:
            n_surv2 += 1
        if local > best:
            best = local
            best_code = rec["code"]
            best_seed = lseed
        # Recorded seed's ACF must match the file (that seed, not necessarily max).
        wrec = sigma.iterate(rec["seed"], 2187)
        Lrec = first_acf_len(wrec)
        if Lrec != rec["acf"]:
            acf_mismatch += 1
            print("FAIL push acf", rec, "python", Lrec)
            fails += 1
    print("python_n_surv_2187", n_surv2)
    print("python_best2_acf", best)
    print("python_best2_code", best_code)
    print("python_best2_seed", best_seed)
    if n_surv2 != 0 or best != 1647 or best_code != 12507201 or best_seed != 1:
        print("FAIL python maximizer among survivors")
        fails += 1
    if acf_mismatch:
        print("FAIL acf_mismatch", acf_mismatch)
        fails += 1

    sigma, imgs = sigma_from_code(12507201)
    print("max_imgs", imgs)
    T = T_from_imgs(imgs)
    print("max_T", T)
    if imgs != [(1, 0, 0), (1, 0, 2), (1, 3, 2), (3, 3, 2)]:
        print("FAIL imgs")
        fails += 1
    if T != (1, 3, 6, 8):
        print("FAIL T")
        fails += 1
    M = incidence_from_imgs(imgs)
    print("max_M", M)
    if not is_primitive(M):
        print("FAIL not primitive")
        fails += 1
    if imgs[1][0] != 1:
        print("FAIL not prolongable at 1")
        fails += 1
    w = sigma.iterate(1, 2187)
    cube = find_cube(w)
    acf = first_acf_len(w)
    print("max_cube", cube)
    print("max_acf", acf)
    if cube != (982, 222, 229) or acf != 1647:
        print("FAIL maximizer cube/acf")
        fails += 1
    if cube and cube[0] + 3 * cube[1] - 1 != 1647:
        print("FAIL cube/ACF identity")
        fails += 1
    # Prefix of length 1647 is ACF; 1648 contains the cube.
    if find_cube(w[:1647]) is not None:
        print("FAIL prefix 1647 has a cube")
        fails += 1
    if find_cube(w[:1648]) is None:
        print("FAIL prefix 1648 has no cube")
        fails += 1

    cpoly = charpoly_coeffs(M)
    print("max_charpoly", cpoly)
    # t^4 - 6 t^3 + 8 t^2 - 3 t  (Faddeev); record for the report.
    # Independently: det(tI-M).
    def det4(A):
        def det3(B):
            a, b, c = B[0]
            d, e, f = B[1]
            g, h, i = B[2]
            return a * (e * i - f * h) - b * (d * i - f * g) + c * (d * h - e * g)

        s = 0
        for j in range(4):
            minor = [[A[r][c] for c in range(4) if c != j] for r in range(1, 4)]
            s += (1 if j % 2 == 0 else -1) * A[0][j] * det3(minor)
        return s

    # Evaluate p(t)=det(tI-M) at a few integers and match Faddeev.
    for tval in (0, 1, 2, 3, 6):
        tIM = [[(tval if i == j else 0) - M[i][j] for j in range(4)] for i in range(4)]
        d = det4(tIM)
        # p(t) = t^4 + c1 t^3 + c2 t^2 + c3 t + c4
        pred = tval**4 + cpoly[1] * tval**3 + cpoly[2] * tval**2 + cpoly[3] * tval + cpoly[4]
        if d != pred:
            print("FAIL charpoly at", tval, d, pred)
            fails += 1
    print("OK charpoly interpolation")

    pref = w[:1647]
    pmax = dmax = 0
    holes_max = 0
    n_with_trip = 0
    for n in range(1, min(80, 1647 // 3) + 1):
        r = additive_at(pref, n)
        pmax = max(pmax, r["p"])
        dmax = max(dmax, r["diam"])
        holes_max = max(holes_max, r["holes"])
        if r["ntrip"]:
            n_with_trip += 1
        if r["p"] > r["diam"] + 1:
            print("FAIL p>diam+1", n, r)
            fails += 1
    print("max_acfpref_pmax", pmax, "diam_max", dmax, "holes_max", holes_max, "n_with_trip", n_with_trip)
    if n_with_trip != 0:
        print("FAIL trips inside ACF prefix")
        fails += 1
    # Growing p_T on the finite ACF prefix (Conjecture A test on a finite word).
    r1 = additive_at(pref, 1)
    r8 = additive_at(pref, 8)
    r16 = additive_at(pref, 16)
    r64 = additive_at(pref, 64)
    print("max_pT", r1["p"], r8["p"], r16["p"], r64["p"])
    print("max_diam", r1["diam"], r8["diam"], r16["diam"], r64["diam"])
    if not (r1["p"] < r8["p"] <= r16["p"] and r16["p"] < r64["p"]):
        # Allow plateaus but require overall growth 1 -> 64.
        if r64["p"] <= r1["p"]:
            print("FAIL p_T not growing on maximizer ACF prefix")
            fails += 1
    # The actual cube length-set contains i, i+n, i+2n.
    full = additive_at(w[:1648], 222)
    if full["ntrip"] < 1:
        print("FAIL cube length 222 has no triple on prefix 1648")
        fails += 1
    else:
        print("OK level set n=222 has i,i+n,i+2n on prefix 1648")

    # Other dual-checked long maps from the census.
    for code, seed, cube_expect, acf_expect in (
        (16376976, 0, (539, 222, 229), 1204),
        (8805231, 1, (409, 144, 235), 840),
        (4269726, 2, (409, 144, 197), 840),
    ):
        sig, _imgs = sigma_from_code(code)
        ww = sig.iterate(seed, 2187)
        cu = find_cube(ww)
        L = first_acf_len(ww)
        print("map", code, "cube", cu, "acf", L)
        if cu != cube_expect or L != acf_expect:
            print("FAIL long map", code)
            fails += 1

    # Controls.
    U = FAMOUS.iterate(2, 256)
    fam_cube = find_cube(U)
    fam_acf = first_acf_len(U)
    print("famous_cube", fam_cube, "acf", fam_acf)
    if fam_cube != (45, 25, 35) or fam_acf != 119:
        print("FAIL famous")
        fails += 1
    fam_pref = U[:119]
    fam_pmax = max(additive_at(fam_pref, n)["p"] for n in range(1, min(40, 119 // 3) + 1))
    print("famous_acfpref_pmax", fam_pmax)
    if fam_pmax < 8:
        print("FAIL famous p_T not visibly growing")
        fails += 1

    cass = cassaigne_iterate(2048)
    if find_cube(cass) is not None:
        print("FAIL Cassaigne cube in 2048")
        fails += 1
    cass_prof = [additive_at(cass, n) for n in (1, 8, 16, 32, 64)]
    cass_pmax = max(additive_at(cass, n)["p"] for n in range(1, 65))
    cass_trips = sum(1 for n in range(1, 65) if additive_at(cass, n)["ntrip"])
    print("cassaigne_pmax64", cass_pmax, "trips", cass_trips, "p_at", [r["p"] for r in cass_prof])
    if cass_trips != 0 or cass_pmax < 40:
        print("FAIL Cassaigne profile")
        fails += 1
    # Holes: unbounded diameter with gaps (p < diam+1).
    cass16 = additive_at(cass, 16)
    if cass16["holes"] <= 0:
        print("FAIL expected Cassaigne holes at n=16")
        fails += 1

    v = vtm(2000)
    vc = find_cube(v)
    print("vtm_cube", vc)
    if vc != (0, 3, 3):
        print("FAIL vtm")
        fails += 1
    vtm_p = max(additive_at(v, n)["p"] for n in range(1, 65))
    print("vtm_pmax", vtm_p)
    if vtm_p != 3:
        print("FAIL vtm bounded p_T")
        fails += 1

    tm = morph_from_words([[0, 1], [1, 2], [2, 0]], (0, 1, 2))
    tmw = tm.iterate(0, 2048)
    tmc = find_cube(tmw)
    print("tmlike_cube", tmc, "acf", first_acf_len(tmw))
    if tmc != (0, 132, 132):
        print("FAIL tm_like")
        fails += 1
    tml_p = max(additive_at(tmw, n)["p"] for n in range(1, 65))
    print("tmlike_pmax", tml_p)
    if tml_p < 10:
        print("FAIL tm_like p_T not growing")
        fails += 1

    sigma014 = morph_from_words([[0, 1], [1, 4], [4, 0]], (0, 1, 4))
    w014 = sigma014.iterate(0, 256)
    c014 = find_cube(w014)
    print("014_cube", c014, "acf", first_acf_len(w014))
    if c014 != (2, 63, 105) or first_acf_len(w014) != 108:
        print("FAIL 01440")
        fails += 1

    # VdW failure: monochromatic 3-APs need not have difference n.
    U512 = FAMOUS.iterate(2, 512)
    v8 = vdW(U512, 8)
    v25 = vdW(U512, 25)
    print("vdW_famous_n8", v8)
    print("vdW_famous_n25", v25)
    if v8["n_starts_with_some_mono_3AP"] <= v8["n_starts_with_diff_exactly_n"]:
        print("FAIL famous VdW n=8 should have extra APs")
        fails += 1
    cass800 = cassaigne_iterate(800)
    vc8 = vdW(cass800, 8)
    print("vdW_cassaigne_n8", vc8)
    if vc8["n_starts_with_diff_exactly_n"] != 0:
        print("FAIL Cassaigne VdW d=n should be 0")
        fails += 1
    if vc8["n_starts_with_some_mono_3AP"] == 0:
        print("FAIL Cassaigne should have some mono 3-APs")
        fails += 1

    # p <= diam+1 on all listed controls.
    for name, ww, nmax in (
        ("cassaigne", cass, 64),
        ("famous_pref", fam_pref, 39),
        ("vtm", v, 64),
        ("tmlike", tmw, 64),
        ("014", w014, 48),
        ("max_pref", pref, 80),
    ):
        bad = 0
        for n in range(1, nmax + 1):
            if n > len(ww):
                break
            r = additive_at(ww, n)
            if r["p"] > r["diam"] + 1:
                bad += 1
        print(name, "p_le_diam_fails", bad)
        if bad:
            fails += 1

    os.makedirs("data", exist_ok=True)
    with open("data/pt_verify_cert.txt", "w") as f:
        f.write(f"fails {fails}\n")
        f.write(f"best2_acf {best}\n")
        f.write(f"best2_code {best_code}\n")
        f.write(f"max_cube {cube}\n")
        f.write(f"max_T {T}\n")
        f.write(f"max_charpoly {cpoly}\n")
        f.write(f"max_acfpref_pmax {pmax}\n")
        f.write(f"famous_acf {fam_acf}\n")
        f.write(f"cassaigne_pmax64 {cass_pmax}\n")
        f.write(f"vtm_pmax {vtm_p}\n")
        f.write(f"vdW_famous_n8_any {v8['n_starts_with_some_mono_3AP']}\n")
        f.write(f"vdW_famous_n8_dn {v8['n_starts_with_diff_exactly_n']}\n")
        f.write(f"vdW_cass_n8_dn {vc8['n_starts_with_diff_exactly_n']}\n")

    print("pt_verify_fails", fails)
    if fails:
        sys.exit(1)


if __name__ == "__main__":
    main()

File: src/pt_verify.cpp
// Independent C++ dual of the m=3 census claims.
// Does NOT re-run the 16,777,216 ACF search. It:
//   1. recounts prolongable / primitive / 3-sum filters (no ACF);
//   2. rebuilds every cap-243 survivor from data/pt_scan3.txt and recomputes
//      ACF prefixes at cap 2187, trying every prolongable seed;
//   3. lex-first cube of the maximizer via find_cube (not only ending-at);
//   4. famous 2-uniform ACF prefix and a few p_T values.
#include "acf.hpp"
#include <array>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <set>
#include <sstream>
#include <string>
#include <vector>
using namespace acf;

static void apply3(const u8 img[4][3], const std::vector<u8>& w, std::vector<u8>& nxt) {
    nxt.clear();
    nxt.reserve(w.size() * 3);
    for (u8 a : w) {
        nxt.push_back(img[a][0]);
        nxt.push_back(img[a][1]);
        nxt.push_back(img[a][2]);
    }
}

static bool primitive4(const u8 img[4][3]) {
    int M[4][4] = {};
    for (int a = 0; a < 4; ++a)
        for (int r = 0; r < 3; ++r) M[img[a][r]][a]++;
    int acc[4][4];
    for (int i = 0; i < 4; ++i)
        for (int j = 0; j < 4; ++j) acc[i][j] = M[i][j];
    for (int p = 0; p < 12; ++p) {
        bool ok = true;
        for (int i = 0; i < 4; ++i)
            for (int j = 0; j < 4; ++j)
                if (acc[i][j] <= 0) ok = false;
        if (ok) return true;
        int nxt[4][4] = {};
        for (int i = 0; i < 4; ++i)
            for (int k = 0; k < 4; ++k)
                for (int j = 0; j < 4; ++j)
                    nxt[i][j] += acc[i][k] * M[k][j];
        for (int i = 0; i < 4; ++i)
            for (int j = 0; j < 4; ++j) acc[i][j] = nxt[i][j];
    }
    return false;
}

static void decode(uint32_t code, u8 img[4][3]) {
    uint32_t t = code;
    for (int a = 0; a < 4; ++a)
        for (int r = 0; r < 3; ++r) {
            img[a][r] = (u8)(t & 3u);
            t >>= 2;
        }
}

static int n_image_sums(const u8 img[4][3]) {
    bool seen[13] = {};
    int nsum = 0;
    for (int a = 0; a < 4; ++a) {
        int x = (int)img[a][0] + img[a][1] + img[a][2];
        if (x < 0 || x > 12) continue;
        if (!seen[x]) {
            seen[x] = true;
            ++nsum;
        }
    }
    return nsum;
}

static int acf_prefix3(const u8 img[4][3], u8 seed, int cap) {
    std::vector<u8> w = {seed};
    std::vector<i64> S = {0, (i64)seed};
    int n = 1;
    std::vector<u8> nxt;
    while (n < cap) {
        apply3(img, w, nxt);
        if ((int)nxt.size() <= n) break;
        for (int i = n; i < (int)nxt.size() && i < cap; ++i) {
            S.push_back(S.back() + nxt[i]);
            int nn = i + 1;
            if (has_cube_ending_at(S, nn)) return nn - 1;
        }
        n = std::min((int)nxt.size(), cap);
        w.swap(nxt);
        if ((int)w.size() > cap) w.resize(cap);
    }
    return n;
}

static std::vector<u8> iterate3(const u8 img[4][3], u8 seed, int cap) {
    std::vector<u8> w = {seed};
    std::vector<u8> nxt;
    while ((int)w.size() < cap) {
        apply3(img, w, nxt);
        if (nxt.size() <= w.size()) break;
        w.swap(nxt);
    }
    if ((int)w.size() > cap) w.resize(cap);
    return w;
}

static int pT_at(const std::vector<u8>& w, int n) {
    if (n <= 0 || n > (int)w.size()) return 0;
    auto S = prefix_sums(w);
    std::set<i64> vals;
    for (int i = 0; i + n <= (int)w.size(); ++i) vals.insert(S[i + n] - S[i]);
    return (int)vals.size();
}

static int diam_at(const std::vector<u8>& w, int n) {
    if (n <= 0 || n > (int)w.size()) return 0;
    auto S = prefix_sums(w);
    i64 L = S[n] - S[0], H = L;
    for (int i = 0; i + n <= (int)w.size(); ++i) {
        i64 v = S[i + n] - S[i];
        if (v < L) L = v;
        if (v > H) H = v;
    }
    return (int)(H - L);
}

int main() {
    int fails = 0;
    const uint32_t N = 1u << 24;
    int n_prol = 0, n_prim = 0, n_3sum_prol = 0, n_tried = 0;
    for (uint32_t code = 0; code < N; ++code) {
        u8 img[4][3];
        decode(code, img);
        bool prol = false;
        for (int a = 0; a < 4; ++a)
            if (img[a][0] == a) prol = true;
        if (!prol) continue;
        ++n_prol;
        int nsum = n_image_sums(img);
        if (nsum >= 3) ++n_3sum_prol;
        bool prim = primitive4(img);
        if (prim) ++n_prim;
        if (nsum >= 3 && prim) ++n_tried;
    }
    std::cout << "filter_n_prol " << n_prol << "\n";
    std::cout << "filter_n_prim " << n_prim << "\n";
    std::cout << "filter_n_3sum_prol " << n_3sum_prol << "\n";
    std::cout << "filter_n_tried " << n_tried << "\n";
    if (n_prol != 11468800) {
        std::cout << "FAIL filter n_prol\n";
        ++fails;
    }
    if (n_prim != 6993126) {
        std::cout << "FAIL filter n_prim\n";
        ++fails;
    }
    if (n_3sum_prol != 9992241) {
        std::cout << "FAIL filter n_3sum_prol\n";
        ++fails;
    }
    if (n_tried != 6116013) {
        std::cout << "FAIL filter n_tried\n";
        ++fails;
    }

    std::ifstream in("data/pt_scan3.txt");
    if (!in) {
        std::cout << "FAIL missing data/pt_scan3.txt\n";
        return 1;
    }
    std::vector<std::pair<int, int>> pushes; // code, recorded seed
    int file_best2 = -1, file_best2_code = -1, file_best2_seed = -1;
    int file_nsurv2187 = -1, file_surv243 = -1;
    std::string line;
    while (std::getline(in, line)) {
        std::istringstream iss(line);
        std::string key;
        iss >> key;
        if (key == "push") {
            int code, seed, acf;
            std::string sseed, sacf;
            iss >> code >> sseed >> seed >> sacf >> acf;
            pushes.push_back({code, seed});
        } else if (key == "n_surv_2187") {
            iss >> file_nsurv2187;
        } else if (key == "best2_acf") {
            iss >> file_best2;
        } else if (key == "best2_code") {
            iss >> file_best2_code;
        } else if (key == "best2_seed") {
            iss >> file_best2_seed;
        } else if (key == "SURV243") {
            iss >> file_surv243;
        }
    }
    std::cout << "file_push_lines " << (int)pushes.size() << "\n";
    if ((int)pushes.size() != 86 || file_surv243 != 86) {
        std::cout << "FAIL survivor count " << pushes.size() << " SURV243 " << file_surv243 << "\n";
        ++fails;
    }

    int best = -1, best_code = -1, best_seed = -1, n_surv2 = 0;
    const int cap2 = 2187;
    for (auto [code, rec_seed] : pushes) {
        u8 img[4][3];
        decode((uint32_t)code, img);
        if (!primitive4(img) || n_image_sums(img) < 3) {
            std::cout << "FAIL survivor not prim/3sum " << code << "\n";
            ++fails;
            continue;
        }
        int local = 0, lseed = -1;
        for (int a = 0; a < 4; ++a) {
            if (img[a][0] != a) continue;
            int L = acf_prefix3(img, (u8)a, cap2);
            if (L > local) {
                local = L;
                lseed = a;
            }
        }
        if (local >= cap2) ++n_surv2;
        if (local > best) {
            best = local;
            best_code = code;
            best_seed = lseed;
        }
        (void)rec_seed;
    }
    std::cout << "recompute_n_surv_2187 " << n_surv2 << "\n";
    std::cout << "recompute_best2_acf " << best << "\n";
    std::cout << "recompute_best2_code " << best_code << "\n";
    std::cout << "recompute_best2_seed " << best_seed << "\n";
    if (n_surv2 != 0) {
        std::cout << "FAIL survivors at 2187\n";
        ++fails;
    }
    if (best != 1647 || best_code != 12507201 || best_seed != 1) {
        std::cout << "FAIL maximizer identity\n";
        ++fails;
    }
    if (file_best2 != 1647 || file_best2_code != 12507201 || file_nsurv2187 != 0) {
        std::cout << "FAIL file maximizer fields\n";
        ++fails;
    }

    // Maximizer lex-first cube on a long enough iterate.
    {
        u8 img[4][3];
        decode(12507201u, img);
        auto w = iterate3(img, 1, 2187);
        auto c = find_cube(w);
        int acf = acf_prefix3(img, 1, 2187);
        std::cout << "max_acf " << acf << "\n";
        if (!c) {
            std::cout << "FAIL maximizer no cube\n";
            ++fails;
        } else {
            std::cout << "max_cube " << c->i << " " << c->d << " " << c->sum << "\n";
            if (c->i != 982 || c->d != 222 || c->sum != 229) {
                std::cout << "FAIL maximizer cube\n";
                ++fails;
            }
            if (c->i + 3 * c->d - 1 != 1647) {
                std::cout << "FAIL cube vs ACF arithmetic\n";
                ++fails;
            }
        }
        if (acf != 1647) {
            std::cout << "FAIL maximizer acf\n";
            ++fails;
        }
        // T and primitivity
        int T[4];
        for (int a = 0; a < 4; ++a) T[a] = img[a][0] + img[a][1] + img[a][2];
        std::cout << "max_T " << T[0] << " " << T[1] << " " << T[2] << " " << T[3] << "\n";
        if (T[0] != 1 || T[1] != 3 || T[2] != 6 || T[3] != 8) {
            std::cout << "FAIL maximizer T\n";
            ++fails;
        }
        if (!primitive4(img) || img[1][0] != 1) {
            std::cout << "FAIL maximizer prim/prol\n";
            ++fails;
        }
        std::cout << "max_imgs";
        for (int a = 0; a < 4; ++a) {
            std::cout << " ";
            for (int r = 0; r < 3; ++r) std::cout << (int)img[a][r];
        }
        std::cout << "\n";
        // p_T on the ACF prefix
        w.resize(1647);
        std::cout << "max_pT_1 " << pT_at(w, 1) << "\n";
        std::cout << "max_pT_8 " << pT_at(w, 8) << "\n";
        std::cout << "max_pT_16 " << pT_at(w, 16) << "\n";
        std::cout << "max_pT_64 " << pT_at(w, 64) << "\n";
        std::cout << "max_diam_64 " << diam_at(w, 64) << "\n";
    }

    // Famous 2-uniform control.
    {
        u8 img[4][2] = {{3, 2}, {3, 1}, {2, 0}, {0, 1}};
        std::vector<u8> w = {2};
        while ((int)w.size() < 256) {
            std::vector<u8> nxt;
            nxt.reserve(w.size() * 2);
            for (u8 a : w) {
                nxt.push_back(img[a][0]);
                nxt.push_back(img[a][1]);
            }
            w.swap(nxt);
        }
        w.resize(256);
        auto c = find_cube(w);
        auto S = prefix_sums(w);
        int acf = 256;
        for (int n = 3; n <= 256; ++n) {
            if (has_cube_ending_at(S, n)) {
                acf = n - 1;
                break;
            }
        }
        std::cout << "famous_acf " << acf << "\n";
        if (!c) {
            std::cout << "FAIL famous no cube\n";
            ++fails;
        } else {
            std::cout << "famous_cube " << c->i << " " << c->d << " " << c->sum << "\n";
            if (c->i != 45 || c->d != 25 || c->sum != 35 || acf != 119) {
                std::cout << "FAIL famous cube/acf\n";
                ++fails;
            }
        }
        w.resize(119);
        std::cout << "famous_acfpref_pT_8 " << pT_at(w, 8) << "\n";
        std::cout << "famous_acfpref_pT_16 " << pT_at(w, 16) << "\n";
        std::cout << "famous_acfpref_diam_16 " << diam_at(w, 16) << "\n";
    }

    std::cout << "pt_verify_fails " << fails << "\n";
    return fails ? 1 : 0;
}
